TypeCobol/Compiler/Parser/CodeElementsParserStep.cs

```diff
@@ -332,6 +333,49 @@ internal static IList<DocumentChange<ICodeElementsLine>> ParseProcessedTokensLin
 
+                            // check if End statement is aligned with the matching opening statement
+                            if (checkEndAlignment && codeElement.Type.ToString().EndsWith("End"))
+                            {
+                                string endStatement = codeElement.Type.ToString();
+                                if (!endStatement.Equals("ProgramEnd", StringComparison.OrdinalIgnoreCase))
+                                {
+                                    string statement = endStatement.Substring(0, endStatement.Length - 3);
+                                    // search code elements from more recent to the less recent codeElementsLinesChanges
+                                    var list = codeElementsLinesChanges?
+                                        .Where(cel => cel.NewLine.CodeElements != null)?
+                                        .OrderByDescending(cel => cel.LineIndex)
+                                        .Select(cel => cel.NewLine.CodeElements);
+                                    if (list != null)
+                                    {
+                                        int level = 1;
+                                        foreach (var codeElements in list)
+                                        {
+                                            for (int i = codeElements.Count - 1; i >= 0; i--)
+                                            {
+                                                string statementName = codeElements[i].Type.ToString();
+                                                if (statementName == endStatement) level++;
+                                                else if (statementName == statement)
+                                                {
+                                                    level--;
+                                                    if (level == 0)
+                                                    {
+                                                        if (codeElements[i].Column != codeElement.Column &&
+                                                            codeElements[i].Line != codeElement.Line)
+                                                        {
+                                                            DiagnosticUtils.AddError(codeElement,
+                                                                "a End statement is not aligned with the matching opening statement",
+                                                                MessageCode.Warning);
+                                                        }
+                                                        break;
+                                                    }
+                                                }
+                                            }
+                                            if (level == 0) break;
+                                        }
+                                    }
```

Comment:
Ok, this looks a bit clunky :-). Why not check a the Node level with a visitor on End nodes ? End nodes are attached to the corresponding opening statement node therefore you could immediately find the corresponding opening statement node by using Parent property and then proceed to check alignment.

Reply:
modified
https://github.com/TypeCobolTeam/TypeCobol/tree/443efbd3422c032e41827e7c89979c9ddfa2bb44

Reply:
Yes this is the good solution. Just one thing @LANOYDO always think what happend if some code comes from a COPY instruction, in this situation comparing line and columns is not enough, I think in this situation we should ignore the test. For testing if a CodeElement is inside a copy you can use for instance endCodeElement.IsInsideCopy()

Reply:
done with parentCodeElement

> Handle the case some code comes from a COPY instruction, in this situation comparing line and columns is not enough, I think in this situation we should ignore the test. For testing if a CodeElement is inside a copy you can use for instance endCodeElement.IsInsideCopy()